### PR TITLE
feat: new forge frontend

### PR DESCRIPTION
**Pull request by daejunpark**
reimplement forge frontend without crytic-compile dependency

### CODE REVIEW

File: src/halmos/__main__.py [hunk 1]
```diff
@@ -819,19 +836,25 @@ def main() -> int:
     # compile
     #
 
-    try:
-        crytic_compile_args, crytic_compile_unknown_args  = crytic_compile_parser.parse_known_args()
+    build_cmd = [
+        'forge', # shutil.which('forge')
```
Comment (daejunpark) [inline src/halmos/__main__.py:840]: not sure if we need `shutil.which('forge')` here, or not, especially in the windows env. what do you think, karma?

File: pyproject.toml [hunk 2]
```diff
@@ -18,7 +18,6 @@ classifiers = [
 ]
 requires-python = ">=3.8"
 dependencies = [
-    "crytic-compile >= 0.3.0, < 0.3.2",
```
Comment (karmacoma-eth) [inline pyproject.toml:21]: feels good man 💖

File: src/halmos/__main__.py (hunk 1 above)
Comment (karmacoma-eth) [reply src/halmos/__main__.py:840]: I don't think so, this should be fine

File: src/halmos/__main__.py [hunk 3]
```diff
@@ -819,19 +836,25 @@ def main() -> int:
...
+        '--root', args.root,
+        '--extra-output', 'storageLayout', 'metadata'
+    ]
 
-        both_unknown = set(halmos_unknown_args) & set(crytic_compile_unknown_args)
-        if both_unknown:
-            print(color_warn(f'error: unrecognized arguments: {" ".join(both_unknown)}'))
-            return 1
+    # run forge without capturing stdout/stderr
+    build_exitcode = subprocess.run(build_cmd).returncode
```
Comment (karmacoma-eth) [inline src/halmos/__main__.py:847]: I think it's nice that we see the compiler output now 👌

```
[⠔] Compiling...
[⠒] Compiling 1 files with 0.5.17
[⠘] Compiling 43 files with 0.8.19
[⠢] Solc 0.5.17 finished in 9.82ms
[⠒] Solc 0.8.19 finished in 3.40s
```

Comment (karmacoma-eth) [reply src/halmos/__main__.py:847]: (and it makes it easier to troubleshoot if there's a compilation error too)

File: src/halmos/__main__.py [hunk 4]
```diff
@@ -819,19 +836,25 @@ def main() -> int:
...
-    except InvalidCompilation as e:
-        print(color_warn(f'Parse error: {e}'))
+    if build_exitcode:
+        print(color_warn(f'build failed: {build_cmd}'))
         return 1
 
+    try:
+        build_out = parse_build_out(args)
+    except Error as err:
+        print(color_warn(f'build output parsing failed: {err}'))
```
Comment (karmacoma-eth) [inline src/halmos/__main__.py:856]: I think there's a problem with the except block

```
    except Error as err:
NameError: name 'Error' is not defined. Did you mean: 'OSError'?
```

Comment (karmacoma-eth) [reply src/halmos/__main__.py:856]:
```suggestion
    except Exception as ex:
        print(color_warn(f'build output parsing failed: {ex}'))
```

Comment (karmacoma-eth) [reply src/halmos/__main__.py]: should we continue after this exception or abort? 

if we continue, it fails with 

```
    for compiler_version in sorted(build_out):
UnboundLocalError: local variable 'build_out' referenced before assignment
```

File: src/halmos/__main__.py [hunk 5]
```diff
@@ -783,6 +762,49 @@ def mk_arrlen(args: argparse.Namespace) -> Dict[str, int]:
     return arrlen
 
 
+def parse_build_out(args: argparse.Namespace) -> Dict:
+    result = {} # compiler version -> source filename -> contract name -> (json, type)
+
+    out_path = os.path.join(args.root, args.forge_build_out)
+    if not os.path.exists(out_path): raise ValueError('not exist', out_path)
```
Comment (karmacoma-eth) [inline src/halmos/__main__.py:769]:
```suggestion
    if not os.path.exists(out_path): raise ValueError(f'{out_path} does not exist')
```

Comment (karmacoma-eth) [reply src/halmos/__main__.py:769]: alternatively, we may want to throw the built-in `FileNotFoundError`?

**Review by karmacoma-eth — DISMISSED** [review]
minor comments otherwise LGTM! 👌

**Review by karmacoma-eth — DISMISSED** [review]

File: src/halmos/__main__.py (hunk 3 above)
Comment (daejunpark) [reply src/halmos/__main__.py:847]: yes, that was the intention.

File: src/halmos/warnings.py [hunk 6]
```diff
@@ -5,6 +5,7 @@
 from .utils import color_warn
 
 logger = logging.getLogger(__name__)
+logging.basicConfig()
```
Comment (daejunpark) [inline src/halmos/warnings.py:8]: karma, this is needed for the basic formatting of logging messages with the prefix of "WARNING:halmos.warnings:".
so far, it was implicitly executed when importing crytic_compile. please let me know if you don't want to keep this.

File: src/halmos/__main__.py (hunk 4 above)
Comment (daejunpark) [reply src/halmos/__main__.py]: good catch, thanks!!

File: src/halmos/warnings.py (hunk 6 above)
Comment (karmacoma-eth) [reply src/halmos/warnings.py:8]: ah I see! no it's great, let's keep it

**Review by karmacoma-eth — APPROVED** [review]

**Pull request merged**